server/src/main/java/io/crate/types/TypeCompatibility.java

```diff
@@ -108,6 +108,12 @@ private static DataType<?> typeCompatibilityForParametrizedType(DataType<?> from
                 } else {
                     return toType;
                 }
+            } else if (fromType.id() == BitStringType.ID && toType.id() == BitStringType.ID) {
+                if (((BitStringType) fromType).length() > ((BitStringType) toType).length()) {
```

Comment:
I don't know if it makes much sense, it's just an idea, and also not to be included in this PR:
How about we have a utility method `length(DataType dataType)` that can get us the length of types that have length and `NULL` or `-1` (if primitive) otherwise. I don't know if it's possible to have some kind of visitor pattern an avoid castings and `id()` checks. Again just an idea.

Reply:
I agree, we should improve this, I'd think more of an interface instead, e.g. `VariableLengthType`..

Reply:
With an iface I think we still cannot avoid casting (to the iface this time) though.

Reply:
I will do smth about that as a follow up

Reply:
`DataType` would already contains `characterMaximumLength`

Reply:
>DataType would already contain characterMaximumLength

thx, utilised this in https://github.com/crate/crate/pull/13425/commits/6c62649127fa9fff5cb6204f80ea2f75a6a602fa